adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/Features.scala

```diff
@@ -0,0 +1,238 @@
+        // commonly used keys in GTF/GFF2, e.g. via Ensembl
+        case "gene_id"       => f.setGeneId(entry._2)
+        case "transcript_id" => f.setTranscriptId(entry._2)
+        case "exon_id"       => f.setExonId(entry._2)
+        // unrecognized key, save to attributes
+        case _               => remaining += entry
+      }
+    )
+
+    // set list fields if non-empty
+    if (!aliases.isEmpty) {
+      f.setAliases(aliases)
+    }
+    if (!notes.isEmpty) {
+      f.setNotes(notes)
+    }
+    if (!parentIds.isEmpty) {
+      f.setParentIds(parentIds)
+    }
+    if (!dbxrefs.isEmpty) {
+      f.setDbxrefs(dbxrefs)
+    }
+    if (!ontologyTerms.isEmpty) {
+      f.setOntologyTerms(ontologyTerms)
+    }
+
+    // set remaining attributes if non-empty;
+    // any duplicate keys are lost at this point, last one in wins
+    if (!remaining.isEmpty) {
+      f.setAttributes(remaining)
+    }
+
+    f
+  }
+
+  /**
+   * Gather values from various feature fields into a sequence of attribute key value pairs.
+   *
+   * @param feature feature to gather values from
+   * @return values from various feature fields gathered into a sequence of attribute key value pairs
+   */
+  def gatherAttributes(feature: Feature): Seq[(String, String)] = {
+    val attrs = new ArrayBuffer[(String, String)]
+
+    // not sure why scala compiler requires this
+    def addBooleanTuple(b: java.lang.Boolean) = {
+      attrs += Tuple2("Is_circular", b.toString)
+    }
+
+    // ... and why Tuple2 is required here
+    Option(feature.getFeatureId).foreach(attrs += Tuple2("ID", _))
+    Option(feature.getName).foreach(attrs += Tuple2("Name", _))
+    Option(feature.getTarget).foreach(attrs += Tuple2("Target", _))
+    Option(feature.getGap).foreach(attrs += Tuple2("Gap", _))
+    Option(feature.getDerivesFrom).foreach(attrs += Tuple2("Derives_from", _))
+    Option(feature.getIsCircular).foreach(addBooleanTuple)
+    Option(feature.getGeneId).foreach(attrs += Tuple2("gene_id", _))
+    Option(feature.getTranscriptId).foreach(attrs += Tuple2("transcript_id", _))
+    Option(feature.getExonId).foreach(attrs += Tuple2("exon_id", _))
+    for (alias <- feature.getAliases) attrs += Tuple2("Alias", alias)
+    for (note <- feature.getNotes) attrs += Tuple2("Note", note)
+    for (parentId <- feature.getParentIds) attrs += Tuple2("Parent", parentId)
+    for (dbxref <- feature.getDbxrefs) attrs += Tuple2("Dbxref", toString(dbxref))
+    for (ontologyTerm <- feature.getOntologyTerms) attrs += Tuple2("Ontology_term", toString(ontologyTerm))
+    attrs ++= feature.getAttributes.toSeq
+  }
+
+  /**
+   * Build a name for the specified feature considering its attributes.  Used when
+   * writing out to lossy feature formats such as BED, NarrowPeak, and IntervalList.
+   *
+   * @param feature feature
+   * @return a name for the specified feature considering its attributes
+   */
+  def nameOf(feature: Feature): String = {
+    Option(feature.getName).foreach(return _)
```

Comment:
this and various permutations doesn't work for me; compile errors or `value getOrElse is not a member of java.io.Serializable`